### PR TITLE
[Operator Mechanism] Add scatter_nd_add double gradient support

### LINKED ISSUE #79493: [Autograd] Missing higher-order gradients block molecular force training
# [Autograd] Missing higher-order gradients block molecular force training

## Environment

- PaddlePaddle: 3.1.0
- Wheel commit: `cdcd76a93d1dfe767dde768198e1d7c7488b93f6`
- Latest `develop` inspected: `f92f172716c6190c36573057cf16750d8b1c07b9`
- Python: 3.10
- Device: CPU
- Platform: Linux x86_64

## Use case

An atomistic model predicts energy `E(R, theta)`, computes forces as
`F = -dE/dR`, and trains against force labels. The force loss differentiates
the coordinate gradient with respect to model parameters, so every operator on
the energy path must support higher-order automatic differentiation.

`paddle.atan2`, `paddle.linalg.cross`, `paddle.nn.Embedding`, and
`paddle.scatter_nd_add` currently prevent this workflow.

## Minimal reproduction

Save the following as `reproduce.py`. Run each case in a separate process
because the `cross` case terminates the interpreter.

```python
import argparse

import paddle


def atan2_case():
    x = paddle.to_tensor([0.4, -0.7], dtype="float32")
    y = paddle.to_tensor([1.2, 0.8], dtype="float32")
    x.stop_gradient = False
    y.stop_gradient = False
    loss = paddle.atan2(x, y).square().sum()
    return paddle.grad(loss, [x, y], create_graph=True)


def cross_case():
    x = paddle.to_tensor(
        [[0.2, 0.4, -0.5], [0.7, -0.3, 0.6]], dtype="float32"
    )
    y = paddle.to_tensor(
        [[-0.1, 0.8, 0.3], [0.5, 0.2, -0.4]], dtype="float32"
    )
    x.stop_gradient = False
    loss = paddle.linalg.cross(x, y).square().sum()
    return paddle.grad(loss, x, create_graph=True)


def embedding_case():
    paddle.seed(42)
    embedding = paddle.nn.Embedding(8, 4)
    linear = paddle.nn.Linear(8, 4)
    atom_ids = paddle.to_tensor([1, 3], dtype="int64")
    coordinates = paddle.to_tensor(
        [[0.2, 0.4, -0.5, 0.1], [0.7, -0.3, 0.6, 0.2]],
        dtype="float32",
    )
    coordinates.stop_gradient = False
    features = paddle.concat([embedding(atom_ids), coordinates], axis=-1)
    energy = linear(features).square().sum()
    return paddle.grad(energy, coordinates, create_graph=True)


def scatter_nd_add_case():
    updates = paddle.to_tensor(
        [[0.2, 0.4], [-0.5, 0.1], [0.7, -0.3]], dtype="float32"
    )
    updates.stop_gradient = False
    index = paddle.to_tensor([[0], [1], [0]], dtype="int64")
    aggregated = paddle.scatter_nd_add(
        paddle.zeros([2, 2], dtype="float32"), index, updates
    )
    loss = aggregated.square().sum()
    return paddle.grad(loss, updates, create_graph=True)


CASES = {
    "atan2": atan2_case,
    "cross": cross_case,
    "embedding": embedding_case,
    "scatter_nd_add": scatter_nd_add_case,
}

parser = argparse.ArgumentParser()
parser.add_argument("case", choices=CASES)
args = parser.parse_args()
paddle.set_device("cpu")
print(f"paddle={paddle.__version__}, device=cpu", flush=True)
print(CASES[args.case](), flush=True)
```

```bash
python reproduce.py atan2
python reproduce.py cross
python reproduce.py embedding
python reproduce.py scatter_nd_add
```

## Actual behavior

| API | Result with `create_graph=True` |
| --- | --- |
| `paddle.atan2` | `atan2_grad` has no grad op |
| `paddle.linalg.cross` | CPU SIGSEGV in `CrossGradKernel` |
| `paddle.nn.Embedding` | `embedding_grad` has no grad op |
| `paddle.scatter_nd_add` | `scatter_nd_add_grad` has no grad op |

The embedding reproducer deliberately joins embedding and coordinate branches
before a nonlinear energy. A disconnected or purely linear lookup does not
exercise the force-training graph.

The current `develop` operator YAML still has first-gradient declarations for
these APIs but no corresponding double-gradient declarations. This statement is
based on source inspection, not an unbuilt development wheel.

## Expected behavior

`paddle.grad(..., create_graph=True)` should construct a differentiable first
gradient so a subsequent force loss can backpropagate to model parameters. An
unsupported higher-order derivative should report an exception and must not
segfault.

## Current workarounds

- Replace `atan2` with differentiable cosine/sine components and a detached
  forward-value correction.
- Expand the cross product into elementwise multiplication and subtraction.
- Replace embedding lookup with one-hot matrix multiplication for force models.
- Replace scatter reduction with one-hot matrix multiplication for force models.

These workarounds match the reference PyTorch forward values and parameter
gradients, but increase memory use and require framework-specific branches.

A focused patch adding `scatter_nd_add_double_grad` and a CPU dygraph unit test
was submitted as [PaddlePaddle/Paddle#79494](https://github.com/PaddlePaddle/Paddle/pull/79494).

### PR DESCRIPTION
**Pull request by leeleolay**
### PR Category

Operator Mechanism

### PR Types

Bug fixes

### Description

Fixes #79493

#### Motivation

Molecular force models compute forces as coordinate gradients of energy and differentiate a force loss again during training. `paddle.scatter_nd_add` currently raises:

```text
The Op scatter_nd_add_grad doesn't have any grad op.
```

when the first gradient is requested with `create_graph=True`.

#### Changes

- Register `scatter_nd_add_double_grad` in `backward.yaml`.
- Implement the double gradient as `grad_x_grad + scatter_nd(index, grad_updates_grad)`.
- Handle either optional incoming second gradient.
- Add `scatter_nd_add_grad` to the PIR VJP blacklist because its double gradient is composite-only, consistent with `index_put_grad` and `put_along_axis_grad`.
- Add CPU dygraph coverage for the default `create_graph=True` path and all three optional-input combinations under eager prim.

#### Validation

- Full CPU `paddle_python` build completed successfully (`1495/1495`).
- `python -m unittest test_scatter_nd_op.TestScatterNdAddDoubleGrad`: 2 tests passed.
- Full `test_scatter_nd_op.py`: 29 tests passed, 4 device-specific tests skipped.
- Pre-commit passed for all four changed files.
- Duplicate-index second gradients are covered for only `grad_x_grad`, only `grad_updates_grad`, and both inputs.

### 是否引起精度变化

否

### CODE REVIEW

File: test/legacy_test/test_scatter_nd_op.py [hunk 1]
```diff
@@ -645,6 +645,37 @@ def test_dygraph(self):
...
+            )
+            index = paddle.to_tensor([[0], [0], [1]], dtype="int64")
+            x.stop_gradient = False
+            updates.stop_gradient = False
+
+            output = paddle.scatter_nd_add(x, index, updates)
+            loss = output.square().sum()
+            x_grad, updates_grad = paddle.grad(
+                loss, [x, updates], create_graph=True
+            )
```
Comment (risemeup1111) [inline test/legacy_test/test_scatter_nd_op.py]: P2 **优先级：P2**
**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

当前用例在第一次 `paddle.grad` 中同时请求 `[x, updates]`，因此后续 double grad 会同时给 `x_grad` 和 `updates_grad` 传入 cotangent；但本 PR 的 YAML/实现新增了 `grad_x_grad`、`grad_updates_grad` 两个 optional 分支，且 #79493 的最小复现只请求 `updates` 的一阶梯度。建议补一个只请求单个一阶梯度输出的断言，避免 optional 输出映射或空分支回归时漏测。

可参考把计算封装成 helper 后增加类似用例：

```python
updates_grad = paddle.grad(loss, updates, create_graph=True)[0]
updates_double_grad = paddle.grad(updates_grad.sum(), updates)[0]
np.testing.assert_allclose(
    updates_double_grad.numpy(),
    np.array([[4.0, 4.0], [4.0, 4.0], [2.0, 2.0]]),
)
```

若认为现有 `[x, updates]` 用例已能覆盖单输出可选分支，请说明依据。

**Review by risemeup1111 — COMMENTED** [review]
未发现需要阻塞合入的问题。关于 optional double-grad 分支的测试覆盖有一条非阻塞建议，详情见行级评论；Pre Commit 已通过，其余 CI 以最终结果为准。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: test/legacy_test/test_scatter_nd_op.py (hunk 1 above)
Comment (leeleolay) [reply test/legacy_test/test_scatter_nd_op.py]: Done. The test now covers only `grad_x_grad`, only `grad_updates_grad`, and both optional inputs with duplicate indices. I also added a separate default-eager assertion for the original `create_graph=True` reproduction. The composite second-gradient numerical checks run with eager prim enabled, consistent with other composite-only higher-gradient tests. Full CPU build and the complete `test_scatter_nd_op.py` pass.

File: test/legacy_test/test_scatter_nd_op.py [hunk 2]
```diff
@@ -645,6 +645,37 @@ def test_dygraph(self):
...
+            x_double_grad, updates_double_grad = paddle.grad(
```
[the remark this answers is not in the document]
Comment (leeleolay) [reply test/legacy_test/test_scatter_nd_op.py]: Done. Added the two single-optional branches and retained the combined branch. The expected values are `[[2, 2], [2, 2]]` / all update rows `[[2, 2], ...]` for only `grad_x_grad`, and `[[4, 4], [2, 2]]` / `[[4, 4], [4, 4], [2, 2]]` for only `grad_updates_grad`. Full CPU build and tests pass.

File: test/legacy_test/test_scatter_nd_op.py (hunk 1 above)
Comment (risemeup1111) [reply test/legacy_test/test_scatter_nd_op.py]: 已确认当前提交已覆盖原建议：`test_create_graph` 保留默认 eager 下的 `create_graph=True` 复现路径，`test_double_grad` 在 eager prim 下分别覆盖 only `grad_x_grad`、only `grad_updates_grad` 和两者同时存在的分支，重复 index 的期望值也对齐。该线程我会标记为已解决。

**Review by risemeup1111 — APPROVED** [review]
已复查当前提交，之前关于 optional 分支测试覆盖的建议已补齐，未发现新的阻塞问题。细节见对应的行级讨论。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by leeleolay** [comment]
The approval gate now passes the review-bot check and reports two remaining code-owner requirements. Could @wanghuancoder please review the `backward.yaml` / `paddle/phi` changes, and could @xiaoguoguo626807 please review the new `composite` registration?

Validation completed locally: full CPU `paddle_python` build 1495/1495, generated PIR VJP targets 80/80, complete `test_scatter_nd_op.py` 29 tests passed (4 skipped), and all pre-commit hooks passed. The new tests cover default eager `create_graph=True` plus only-x, only-updates, and combined optional double-gradient branches. Additional `paddle/phi` gate checks on the same build also pass: zero-size trailing dimensions (`out`, x double grad, and updates double grad all shape `[2, 0]`) and analytical precision checks for both float32 and float64. BigTensor does not exercise new code here because this PR changes no kernel or index arithmetic; the composite double gradient delegates to the existing `scatter_nd_add` implementation.

**Comment by leeleolay** [comment]
CI status note: the only non-successful job, Mac-CPU / Build and test, was cancelled by the workflow's 60-minute job timeout. The macOS build completed successfully, and 357/1952 tests had passed with no test failure when GitHub emitted 'The operation was canceled' exactly at the timeout. Coverage, CI-Windows, CI-Build, and CI-H all passed. Please rerun the Mac job if a green aggregate CI result is required.
